### ISSUE TITLE
tableMap codec with list value is irreversible

### ISSUE THREAD

**Issue by samhh** [kowainik/tomland#335]
Me again! :wave: 

It's possible I'm misusing the codecs. I've tried and failed to simplify the repro beyond this point, hopefully this helps.

```haskell
import qualified Data.Map as M
import qualified Toml as Toml
import Toml ((.=))

data Inner = Inner { y :: Text } deriving (Show)
innerCodec = Inner <$> Toml.text "y" .= y

data Thing = Thing { xs :: Map Text [Inner] } deriving (Show)
thingCodec = Thing <$> Toml.tableMap Toml._KeyText (Toml.list innerCodec) "xs" .= xs

thing = Thing $ M.fromList [("k", [Inner "a", Inner "b"])]

reversed = Toml.decode thingCodec $ Toml.encode thingCodec thing
-- Right (Thing {xs = fromList []})
```

The encoded value looks like this written to a file with `Toml.encodeToFile`:

```toml

[xs]
  [[xs.k]]
    y = "a"

  [[xs.k]]
    y = "b"
```

**Comment by chshersh** [comment]
Hi @samhh, can you also tell what output you expect from the codec? It's not clear what's the problem.
Also, generally, specifying top-level signatures of all codecs can help with resolving issues, because due to type inference the inferred types of codecs might not be the same you expect.

**Comment by samhh** [comment]
Hey @chshersh,

In my real codebase I do have explicit type signatures, just tried to write a pared down repro. :slightly_smiling_face: 

What's confusing me is that the encoded value cannot be decoded - is this expected behaviour? I'd generally expect `decode (encode x) == x` to hold.

**Comment by chshersh** [comment]
@samhh Thanks for more details! From the implementation of the `tableMap` codec I can see that it doesn't handle array of tables as an argument codec to `tableMap` at the moment. I guess, it should be possible to patch `tableMap` and take this into consideration.
Just wonder, what is your real use case and why you want to have such nested configurations? I'm curious about different use-cases for TOML and how it can be used in various scenarios 🙂

**Comment by samhh** [comment]
My use case is [something I've just started working on](https://github.com/samhh/terpod).

I'm using TOML for user configuration and, until I'm given a compelling reason not to, also for data caching. The structure of said data is currently like this:

```haskell
data Cache = Cache
  { timestamp :: LocalTime,
    feeds :: Map PodcastId (Map EpisodeId Episode)
  }
```

I'm contemplating changing that to this, which prompted this ticket:

```haskell
data Cache = Cache
  { timestamp :: LocalTime,
    feeds :: Map PodcastId [Episode]
  }
```

**Comment by vrom911** [comment]
@samhh we prepared a fix in #337 . Let us know if it doesn't work for your use-case 🙂 

Current known limitation of the solution – it doesn't preserve empty lists in values of the map.

**Comment by samhh** [comment]
That's working perfectly, many thanks!
